keps/sig-node/20200110-apparmor-ga.md

```diff
@@ -0,0 +1,387 @@
+    ...
+    // The AppArmor options to use by the containers in this pod.
+    // +optional
+    AppArmor  *AppArmorProfile
+    ...
+}
+
+type SecurityContext struct {
+    ...
+    // The AppArmor options to use by this container. If AppArmor options are
+    // provided at both the pod & container level, the container options
+    // override the pod options.
+    // +optional
+    AppArmor  *AppArmorProfile
+    ...
+}
+
+type AppArmorProfileType string
+
+const (
+    AppArmorProfileUnconfined AppArmorProfileType = "Unconfined"
+    AppArmorProfileDefault    AppArmorProfileType = "Default"
+    AppArmorProfileLocalhost  AppArmorProfileType = "Localhost"
+)
+
+// Only one profile source may be set.
+// +union
+type AppArmorProfile struct {
+    // +unionDescriminator
+    Type AppArmorProfileType
+
+    // Load a profile defined on the node.
+    // The profile must be available on the node to work.
+    // The length of the profile is limited to 253 characters.
+    // The `Type` in this struct has to be set to `AppArmorProfileLocalhost`.
+    // +optional
+    LocalhostProfile *string
+}
+```
+
+This API makes the options more explicit and leaves room for new profile sources
+to be added in the future (e.g. Kubernetes predefined profiles or ConfigMap
+profiles). The AppArmor options structure leaves room for future extensions,
+such as defining the behavior when a profile cannot be set.
+
+#### PodSecurityPolicy API
+
+```go
+type PodSecurityPolicySpec struct {
+    ...
+    // AppArmor is the strategy that will dictate allowable and default AppArmor
+    // profiles for the container.
+    // +optional
+    AppArmor *ApparmorStrategyOptions
+    ...
+}
+
+type AppArmorStrategyOptions struct {
+    // The default profile to set on the pod, if none is specified.
+    // The default MUST be allowed by the allowedProfiles.
+    // +optional
+    DefaultProfile *v1.AppArmorProfile
+
+    // The set of profiles that may be set on the pod or containers.
+    // If unspecified, AppArmor profiles are unrestricted by this policy.
+    // +optional
+    AllowedProfiles *AppArmorProfileSet
+}
+
+// A set of AppArmor profiles. This struct should be a plural of
+// `v1.AppArmorProfile`.
+// All values are optional, and an unspecified field excludes all profiles of
+// that type from the set.
+type AppArmorProfileSet struct {
+    // The allowed AppArmor profile types.
+    // +optional
+    Types []AppArmorProfileType
+
+    // The allowed runtimeProfiles. A value of '*' allows all runtimeProfiles.
+    // +optional
+    RuntimeProfiles []string
+
+    // The allowed localhostProfiles. Values may end in '*' to include all
+    // localhostProfiles with a prefix.
+    // +optional
+    LocalhostProfiles []string
+}
+```
+
+## Design Details
+
+### Version Skew Strategy
+
+All API skew is resolved in the API server. New Kubelets will only use the
+AppArmor values specified in the fields and ignore the annotations therefore.
```

Comment:
It may be worth reconsidering this. For seccomp the kubelet supports both, however, it prioritises fields over annotations. On that case, such support will then be dropped at v1.23.

If you were to change this, I would make sure you make that priority order explicit on this KEP.

Reply:
Okay, changed the section to sync up with the seccomp KEP.